libbpf-tools/capable.c

```diff
@@ -0,0 +1,411 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Based on capable(8) from BCC by Brendan Gregg.
+//
+// Copyright 2022 Sony Group Corporation
+
+#include <argp.h>
+#include <signal.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <time.h>
+#include <bpf/bpf.h>
+#include "capable.h"
+#include "capable.skel.h"
+#include "trace_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+
+static struct env {
+	bool	verbose;
+	char	*cgroupspath;
+	bool	cg;
+	bool	extra_fields;
+	bool	user_stack;
+	bool	kernel_stack;
+	bool	unique;
+	char	*unique_type;
+	int	stack_storage_size;
+	int	perf_max_stack_depth;
+	pid_t	pid;
+} env = {
+	.pid = -1,
+	.stack_storage_size = 1024,
+	.perf_max_stack_depth = 127,
+	.unique = false,
+};
+
+const char *cap_name[] = {
+	[0] = "CAP_CHOWN",
+	[1] = "CAP_DAC_OVERRIDE",
+	[2] = "CAP_DAC_READ_SEARCH",
+	[3] = "CAP_FOWNER",
+	[4] = "CAP_FSETID",
+	[5] = "CAP_KILL",
+	[6] = "CAP_SETGID",
+	[7] = "CAP_SETUID",
+	[8] = "CAP_SETPCAP",
+	[9] = "CAP_LINUX_IMMUTABLE",
+	[10] = "CAP_NET_BIND_SERVICE",
+	[11] = "CAP_NET_BROADCAST",
+	[12] = "CAP_NET_ADMIN",
+	[13] = "CAP_NET_RAW",
+	[14] = "CAP_IPC_LOCK",
+	[15] = "CAP_IPC_OWNER",
+	[16] = "CAP_SYS_MODULE",
+	[17] = "CAP_SYS_RAWIO",
+	[18] = "CAP_SYS_CHROOT",
+	[19] = "CAP_SYS_PTRACE",
+	[20] = "CAP_SYS_PACCT",
+	[21] = "CAP_SYS_ADMIN",
+	[22] = "CAP_SYS_BOOT",
+	[23] = "CAP_SYS_NICE",
+	[24] = "CAP_SYS_RESOURCE",
+	[25] = "CAP_SYS_TIME",
+	[26] = "CAP_SYS_TTY_CONFIG",
+	[27] = "CAP_MKNOD",
+	[28] = "CAP_LEASE",
+	[29] = "CAP_AUDIT_WRITE",
+	[30] = "CAP_AUDIT_CONTROL",
+	[31] = "CAP_SETFCAP",
+	[32] = "CAP_MAC_OVERRIDE",
+	[33] = "CAP_MAC_ADMIN",
+	[34] = "CAP_SYSLOG",
+	[35] = "CAP_WAKE_ALARM",
+	[36] = "CAP_BLOCK_SUSPEND",
+	[37] = "CAP_AUDIT_READ",
+	[38] = "CAP_PERFMON",
+	[39] = "CAP_BPF",
+	[40] = "CAP_CHECKPOINT_RESTORE"
+};
+
+static volatile sig_atomic_t exiting = 0;
+struct capable_bpf *obj;
```

Comment:
Why not declare this obj inside main

Reply:
Yes, `obj` is not required for `print_map`.
Moved `obj` inside main.